File: src/main/java/com/oycm/algorithm/a/advance/Solution_13.scala
package com.oycm.algorithm.a.advance

object Solution_13 {

  /**
   * 30. 串联所有单词的子串
   * https://leetcode.cn/problems/substring-with-concatenation-of-all-words/description/
   *
   * 串联子串：包含  words 中所有字符串以任意顺序排列连接起来的子串。
   *
   * @param s     字符串
   * @param words 长度相同的字符串数组
   * @return 所有串联子串在 s 中的开始索引
   */
  def findSubstring(s: String, words: Array[String]): List[Int] = {
    /*
    记 words 中字符串长度为 m, 数组长度为 k, t = m * k
    s 长度记为 n，把 s 分割为 长为 m 的 k 个子串，如果子串的单词数和 words 中的单词数相同，则 i-t 就是开始索引的下标
    [0, m),[m, 2m)...,[(k-1)m, km)
    [1, m+1),[m+1, 2m+1)...,[(k-1)m+1, km+1)
    ...
    [n-km, n-km+m)...[n-m,n)
    s 移动一次，需要 t 次找出自己的 分割子串，和 words 进行比较次数

    时间复杂度 O(k + k * n)
    空间复杂度 O(k)
    */
    val n = s.length

    val k = words.length
    val m = words(0).length
    val t = k * m
    val map = scala.collection.mutable.Map[String, Int]()
    var cnt = 0

    val ans = scala.collection.mutable.ArrayBuffer[Int]()
    for (word <- words) {
      map(word) = map.getOrElse(word, 0) + 1
      if (map(word) == 1) {
        cnt += 1
      }
    }
    for (i <- 0 to n - t) {
      val mapT = map.clone()
      var cntT = cnt
      // t 次循环查找字符数
      for (j <- i until i + t by m) {
        val word = s.substring(j, j + m)
        if (mapT.contains(word)) {
          mapT(word) -= 1
          if (mapT(word) == 0) {
            cntT -= 1
          }
        }
      }
      if (cntT == 0) {
        ans.append(i)
      }

    }
    ans.toList
  }

  def main(args: Array[String]): Unit = {
    println(findSubstring("barfoothefoobarman", Array("foo", "bar")))
    println(findSubstring("wordgoodgoodgoodbestword", Array("word", "good", "best", "word")))
    println(findSubstring("barfoofoobarthefoobarman", Array("bar", "foo", "the")))
  }
}

File: src/main/java/com/oycm/month2026/june/Solution_13.java
package com.oycm.month2026.june;

public class Solution_13 {

    /**
     * 3838. <a href="https://leetcode.cn/problems/weighted-word-mapping/description/">带权单词映射</a> 1241
     *
     * @param words
     * @param weights
     * @return
     */
    public String mapWordWeights(String[] words, int[] weights) {
        /*
        给你一个字符串数组 words，其中每个字符串表示一个由小写英文字母组成的单词。
        同时给你一个长度为 26 的整数数组 weights，其中 weights[i] 表示第 i 个小写英文字母的权重。
        单词的 权重 定义为其所有字符权重的 总和。
        对于每个单词，将其权重对 26 取模，并将结果按字母倒序映射到一个小写英文字母（0 -> 'z', 1 -> 'y', ..., 25 -> 'a'）。
        返回一个由所有单词映射后的字符按顺序连接而成的字符串。
         */
        /*
        1 <= words.length <= 100
        1 <= words[i].length <= 10
        weights.length == 26
        1 <= weights[i] <= 100
        words[i] 仅由小写英文字母组成。
         */
        char[] ans = new char[words.length];
        for (int i = 0; i < words.length; i++) {
            int weight = 0;
            for (char c : words[i].toCharArray()) {
                weight += weights[c - 'a'];
            }
            ans[i] = (char) ('z' - weight % 26);
        }

        return new String(ans);
    }

}
